service/history/workflow/update/registry.go

```diff
@@ -215,12 +219,14 @@ func (r *RegistryImpl) Len() int {
 	return len(r.updates)
 }
 
-func (r *RegistryImpl) remover(id string) func() {
-	return func() {
-		r.mu.Lock()
-		defer r.mu.Unlock()
-		delete(r.updates, id)
-	}
+func (r *RegistryImpl) remover(id string) updateOpt {
+	return withCompletionCallback(
+		func() {
+			r.mu.Lock()
+			defer r.mu.Unlock()
```

Comment:
Locking the registry in a callback seems dangerous to me because you have to ensure that whoever invokes the callback doesn't lock as well. It's much simpler if we just follow the pattern where we lock / defer Unlock at the start of any instance method that needs synchronization, and only there.

Reply:
Can you suggest how the registry map could be locked from the Update? Note that everything here is unexported now so whatever happens is what happens inside this package.

There are other approaches to cleaning up completed state machines, but the ones I could come up with are all _worse_ than this.

Reply:
It's just a gotcha for a future dev--not saying the bug exists now

Reply:
Oh sure. I don't like this callback very much. It's just that I haven't been able to come up with anything better, especially when you consider Registry.admit